lsp/nls/src/requests/completion.rs

```diff
@@ -21,42 +26,151 @@ pub fn handle_completion(
         .id_of(params.text_document_position.text_document.uri.as_str())
         .unwrap();
 
+    let text = server.cache.files().source(file_id);
     let start = position_to_byte_index(
         server.cache.files(),
         file_id,
         &params.text_document_position.position,
     )
     .unwrap();
 
+    // This is the environment containing stdlib stuff.
+    // stdlib terms should be loaded here
+    let _env = &server.initial_ctxt.type_env;
+
+    // lin_env is a mapping from idents to IDs.
+    // This is from the Linearizer used to build this linearization.
+    // idealy we should get the id and use it instead of comparing strings
+    let Completed { lin_env: _, .. } = &server.lin_cache.get(&file_id).unwrap();
+
     let locator = (file_id, ByteIndex(start as u32));
     let linearization = server.lin_cache_get(&file_id)?;
 
     Trace::enrich(&id, linearization);
 
-    let item = linearization.item_at(&locator);
+    let trigger = match params.context {
+        Some(CompletionContext {
+            trigger_character: Some(s),
+            ..
+        }) => Some(s),
+        _ => None,
+    };
 
-    if item == None {
-        server.reply(Response::new_ok(id, Value::Null));
-        return Ok(());
+    /// Extract identifiers from a row type which forms a record.
+    /// Ensure that ty is really a row type.
+    fn extract_ident(ty: &Box<Types>) -> Vec<Ident> {
+        ty.iter_as_rows()
+            .filter_map(|item| match item {
+                RowIteratorItem::Row(ident, _) => Some(ident.clone()),
+                RowIteratorItem::Tail(..) => None,
+            })
+            .collect::<Vec<_>>()
     }
 
-    let item = item.unwrap().to_owned();
+    let in_scope = match trigger {
+        Some(s) if s == "." => {
+            // Get the ID before the .
+            // if the current source is `let var = {a = 10} in var.`
+            // name = var
+            // TODO: handle nested record completion from here
+            let name = text[..start - 1]
+                .chars()
+                .rev()
+                .take_while(|c| c.is_ascii_alphabetic() || *c == '.')
+                .skip_while(|c| *c == '.') // skip . (if any)
+                .collect::<String>()
+                .chars()
+                .rev()
+                .collect::<String>();
+
+            linearization
+                .linearization
+                .iter()
```

Comment:
Does this mean we are iterating through the whole linearization each time we want to trigger completion? The linearization can be quite big for big files.

Also, what if the there is another identifier in scope? For example, if I do:

```
let x = {foo = 1} in
let x = {bar = 1} in
x.
```

I get `foo` and `bar` as suggestions, but `foo` will cause an error at runtime.

Reply:
This is kinda tricky. I have noticed that the parser is a bit tolerant some cases, and this is one of such cases. When we type this: `x.` , the parser returns a `ParseError` node. This node might not have the scope of the previous identifiers. Usually , we'd do this `linearization.get_in_scope(&item)` to get the scope, where `item` would be a linearization item that represents the parse error node, but it won't return anything in scope. So searching the whole linearization is a fix  that works for now. 

The second problem can be fixed by checking against unique ids instead of identifier labels, I'll fix it.

Reply:
Hm, do you know why it doesn't return anything in scope? Is it because `ParseError` is treated specifically? My hunch with this PR's solution is that it adds a linear cost in the size of the AST for something that is automatically triggered. If record completion was something that needed an explicit action from the user, then that's fine if it starts slow, it's an additional feature. We can polish it on the way. But this is not the case here, as you get completion as soon as you type a `.`. We tried on very small examples but I wonder how that goes on something like the mantis example, or even bigger.

If the scope solution is complex, then we can go with the current solution for now, but I just want to make sure and to understand why the scope way is _actually_ hard.

Reply:
I'm looking into it

Reply:
We use IDs in scope for completion now, except when we want to check for `RecordFields { .. }`, because those `LinearizationItem`s have their own special scope (only visible to the themselves)